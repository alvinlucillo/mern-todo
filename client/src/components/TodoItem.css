.todo__completed {
  text-decoration: line-through;
}

.todo-item__read-mode {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.todo-item__read-mode > .todo-item__edit {
  margin-left: auto;
}
