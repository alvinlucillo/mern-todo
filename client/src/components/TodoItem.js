import { useState } from "react";

import {
  formatToISODate,
  formatFromISODate,
  toLocaleString,
} from "../utils/Utils";

import {
  ListItem,
  ListItemIcon,
  Checkbox,
  ListItemText,
  ListItemSecondaryAction,
  IconButton,
  TextField,
} from "@mui/material";

import { CalendarToday, Delete, Edit, Done } from "@mui/icons-material";

const TodoItem = ({ todo, onEditMode, onSave, onDelete, onCheck }) => {
  const [title, setTitle] = useState(todo.title);
  const [due, setDue] = useState(formatFromISODate(todo.due));

  const isChecked = todo.status === "completed" ? true : false;

  const handleInputChange = (e) => {
    if (e.target.name === "title") {
      setTitle(e.target.value);
    }

    if (e.target.name === "due") {
      setDue(e.target.value);
    }
  };

  const handleDone = () => {
    onSave({ ...todo, title, due: formatToISODate(due) });
    onEditMode(todo, false);
  };

  const secondaryText = !!todo.isEditMode ? null : (
    <>
      <CalendarToday style={{ fontSize: 12 }} />{" "}
      <span style={{ fontSize: 13 }} className="todoItem__due">
        {toLocaleString(todo.due)}
      </span>
    </>
  );

  const primaryText = !!todo.isEditMode ? (
    <>
      <TextField
        value={title}
        size="small"
        name="title"
        onChange={handleInputChange}
        className="todoItem__edit-textbox-title"
      />{" "}
      <br />
      <TextField
        type="date"
        value={due}
        size="small"
        name="due"
        onChange={handleInputChange}
        margin="dense"
      />
    </>
  ) : (
    <span
      className="todoItem__title"
      style={{ textDecoration: `${isChecked ? "line-through" : "none"}` }}
    >
      {todo.title}
    </span>
  );

  return (
    <ListItem key={todo.id}>
      <ListItemIcon>
        <Checkbox
          onChange={() => onCheck(todo)}
          checked={isChecked}
          className="todoItem__status"
        />
      </ListItemIcon>
      <ListItemText primary={primaryText} secondary={secondaryText} />
      <ListItemSecondaryAction>
        {!!todo.isEditMode === false && (
          <>
            <IconButton
              onClick={() => onEditMode(todo, true)}
              className="todoItem__edit-button"
            >
              <Edit />
            </IconButton>
            <IconButton onClick={() => onDelete(todo)}>
              <Delete />
            </IconButton>
          </>
        )}

        {!!todo.isEditMode && (
          <>
            <IconButton onClick={handleDone}>
              <Done />
            </IconButton>
          </>
        )}
      </ListItemSecondaryAction>
    </ListItem>
  );
};

export default TodoItem;
