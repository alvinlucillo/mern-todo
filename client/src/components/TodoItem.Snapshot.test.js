import React from "react";
import renderer from "react-test-renderer";
import TodoItem from "./TodoItem";
import mockData from "../mocks/data";

describe("TodoItem", () => {
  it("should run successfully", () => {
    const todo = mockData.todos[0];
    const [onEditMode, onSave, onDelete, onCheck] = Array(4).fill(jest.fn());
    const tree = renderer
      .create(
        <TodoItem
          todo={todo}
          onEditMode={onEditMode}
          onSave={onSave}
          onDelete={onDelete}
          onCheck={onCheck}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
