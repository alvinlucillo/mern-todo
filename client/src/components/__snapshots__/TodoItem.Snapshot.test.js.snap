// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should run successfully 1`] = `
<li
  className="MuiListItem-container css-oyakqy-MuiListItem-container"
>
  <div
    className="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-secondaryAction css-hcmkxp-MuiListItem-root"
    disabled={false}
  >
    <div
      className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
    >
      <span
        className="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root todoItem__status css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <input
          checked={false}
          className="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate={false}
          onChange={[Function]}
          type="checkbox"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
    </div>
    <div
      className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
      >
        <span
          className="todoItem__title"
          style={
            Object {
              "textDecoration": "none",
            }
          }
        >
          Todo 1
        </span>
      </span>
      <p
        className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CalendarTodayIcon"
          focusable="false"
          style={
            Object {
              "fontSize": 12,
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"
          />
        </svg>
         
        <span
          className="todoItem__due"
          style={
            Object {
              "fontSize": 13,
            }
          }
        >
          Sunday, September 26, 2021
        </span>
      </p>
    </div>
  </div>
  <div
    className="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium todoItem__edit-button css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="EditIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </button>
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
    </button>
  </div>
</li>
`;
